### PR TITLE
Implemented ad hoc generateTag to avoid buffer copies and reduce garbage produced

### PR DESCRIPTION
**Pull request by franz1981**
The provided implementation is meant to improve throughput and reduce garbage
allocation by using a zero-copy implementation of positive integer ascii encoding.

### CODE REVIEW

**Comment by vietj** [comment]
can you provide functional tests for these methods to avoid regressions ?

**Comment by franz1981** [comment]
Yes, sure and I was thinking to give a better home for that too, as a proper util class, wdyt?

**Comment by vietj** [comment]
that sounds perfect

**Comment by franz1981** [comment]
Perfect!There is any preferred package where I can put that util class?

**Comment by vietj** [comment]
in an impl package

**Comment by gemmellr** [comment]
This change appears like it would actually break the delivery tags. In particular, it would mishandle wraparound of the int and then create the same delivery tag for every 10th message.

**Comment by franz1981** [comment]
@gemmellr I have written a naive test covering 0 -> 100_000_000 produced tags and it gives the same result of the original version.
 The only limitation is that it just deals with positive numbers, but could be changed with ease to cover negative ones too.
Am I missing somenthing?

**Comment by franz1981** [comment]
@vietj @gemmellr 
I have added both a test and a proper util class: i do not love the naming I have used and the test is just an attempt to check correctness while sacrificing reproducibility, but I hope is enough..
:+1:

File: src/main/java/io/vertx/proton/impl/ProtonSenderImpl.java [hunk 1]
```diff
@@ -57,7 +54,7 @@ public ProtonDelivery send(Message message, Handler<ProtonDelivery> onUpdated) {
   }
 
   private byte[] generateTag() {
-    return String.valueOf(tag++).getBytes(StandardCharsets.UTF_8);
+    return IntegerUtil.positiveToAscii(tag++);
```
Comment (franz1981) [inline src/main/java/io/vertx/proton/impl/ProtonSenderImpl.java]: I don't know if tag could be > `Integer.MAX_VALUE`

**Comment by gemmellr** [comment]
> @gemmellr I have written a naive test covering 0 -> 100_000_000 produced tags and it gives the same result of the original version.
The only limitation is that it just deals with positive numbers, but could be changed with ease to cover negative ones too.
Am I missing something?

Yes, you are missing that the number provided can wrap around as mentioned and will then become negative, after which it will incorrectly return repeating tags for every 10th message. I had only looked at it when commenting earlier, but I tried it after by manipulating the tag start point.

If the new "positiveToAscii" cant tolerate negatives it shouldnt accept them, but does then gives the wrong answer.

**Comment by franz1981** [comment]
@gemmellr I can change the function in order to check for negative value and throw an exception or make it able to deal with negative ones too. I can't say which one is a preferred approach, any idea?
Another q: the tag could be > Integer.MAX_VALUE? I don't know it...

File: src/main/java/io/vertx/proton/impl/IntegerUtil.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2016, 2017 the original author or authors.
```
Comment (gemmellr) [inline src/main/java/io/vertx/proton/impl/IntegerUtil.java]: Wrong dates

File: src/test/java/io/vertx/proton/impl/IntegerUtilTest.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright 2016 the original author or authors.
```
Comment (gemmellr) [inline src/test/java/io/vertx/proton/impl/IntegerUtilTest.java]: Wrong date.

**Comment by franz1981** [comment]
@gemmellr 
> It doesn't need to be ascii at all, and elsewhere we've just output a raw number using the length of bytes needed to hold it.

So it could use just a byte[4] putting into it the tag as binary?
If the answer is yes IMO it would be even better than this PR, making the code simpler and without implementing a (although cool :P) custom, but useless zero-copy ascii encoding as I have done

**Comment by gemmellr** [comment]
Yes

**Comment by franz1981** [comment]
I would raise another PR doing just the binary version then, much much better IMO!
if @vietj need this "int2Ascii" encoder feel free to use it, I have tested it that vs the jdk version is quite faster  ie 4X!!!

**Comment by vietj** [comment]
@franz1981 I'll keep that on top of my mind whenever I need it

**Pull request closed without merging**